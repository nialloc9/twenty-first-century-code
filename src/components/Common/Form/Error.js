import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import Message from "../Styled/Message";
import { remCalc } from '../../../common/helpers';

class Error extends PureComponent {
    static propTypes = {
        error: PropTypes.string,
        touched: PropTypes.bool
    };

    static defaultProps = {
        error: "",
        touched: true
    };

    render() {
        const { error, touched, ...rest } = this.props;

        if (error === "" || !touched) {
            return null;
        }

        return <Message margin={`${remCalc(20)} auto auto auto`} negative {...rest}>{error}</Message>;
    }
}

export default Error;
