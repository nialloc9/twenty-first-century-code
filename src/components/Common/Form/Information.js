import React, { Component } from "react";
import PropTypes from "prop-types";
import Message from "semantic-ui-react/dist/commonjs/collections/Message";

class Information extends Component {
    static propTypes = {
        information: PropTypes.string,
        touched: PropTypes.bool
    };

    render() {
        const { information, touched } = this.props;

        if (!information || !touched) {
            return null;
        }

        return <Message info>{information}</Message>;
    }
}

export default Information;
