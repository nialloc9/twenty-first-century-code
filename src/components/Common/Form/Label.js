import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import Styled from "styled-components";
import { remCalc } from "../../../common/helpers";

const StyledLabel = Styled("label")`
    float: ${({ float = "left" }) => float};
    margin-left: ${({ marginLeft = remCalc(6) }) => marginLeft};
`;

class Label extends PureComponent {
    static propTypes = {
        inputLabel: PropTypes.string
    };

    render() {
        const { inputLabel, ...rest } = this.props;

        if (!inputLabel) {
            return null;
        }

        return <StyledLabel {...rest}>{inputLabel}</StyledLabel>;
    }
}

export default Label;
