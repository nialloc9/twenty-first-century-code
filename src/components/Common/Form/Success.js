import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import Message from "../Styled/Message";
import { remCalc } from '../../../common/helpers';

class Success extends PureComponent {
    static propTypes = {
        success: PropTypes.string
    };

    static defaultProps = {
        success: ""
    };

    render() {
        const { success, ...rest } = this.props;

        if (success === "") {
            return null;
        }

        return <Message margin={`${remCalc(20)} auto auto auto`} positive {...rest}>{success}</Message>;
    }
}

export default Success;
