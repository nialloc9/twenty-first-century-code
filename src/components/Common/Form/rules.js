import moment from "moment";

export const required = value => (value ? undefined : "Required");

export const minLength = min => value =>
    value && value.length < min
        ? `Must be ${min} characters or more`
        : undefined;

export const maxLength = max => value =>
    value && value.length > max
        ? `Must be ${max} characters or less`
        : undefined;

export const number = value =>
    value && isNaN(Number(value)) ? "Must be a number" : undefined;

export const minValue = min => value =>
    value && value < min ? `Must be at least ${min}` : undefined;

export const maxValue = max => value =>
    value && value < max ? `Must be at most ${max}` : undefined;

export const email = value =>
    value && !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(value)
        ? "Invalid email address"
        : undefined;

export const isValidDate = value => {
    return !moment(value, "DD-MM-YYYY").isValid()
        ? "Must be DD-MM-YYYY"
        : undefined;
};

export const isWholeNumber = value =>
    value && value % 1 === 0 ? undefined : "Must be a whole number";

export default {
    required,
    maxLength,
    number,
    minValue,
    maxValue,
    email,
    isValidDate,
    isWholeNumber
};
