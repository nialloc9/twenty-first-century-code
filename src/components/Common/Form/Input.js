import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import Form from "semantic-ui-react/dist/commonjs/collections/Form";
import StyledInput from "../Styled/Input";
import Label from "./Label";
import Error from "./Error";
import Warning from "./Warning";

class Input extends PureComponent {
    static propTypes = {
        label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        placeholder: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        type: PropTypes.string,
        size: PropTypes.string,
        input: PropTypes.object.isRequired,
        meta: PropTypes.object.isRequired
    };

    render() {
        const { label, meta, input, ...rest } = this.props;

        return (
            <Form.Field>
                <Label inputLabel={label} />
                <StyledInput {...input} {...rest} />
                <Error {...meta} />
                <Warning {...meta} />
            </Form.Field>
        );
    }
}

export default Input;
