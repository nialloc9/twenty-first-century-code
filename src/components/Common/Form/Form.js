import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import SemanticForm from "semantic-ui-react/dist/commonjs/collections/Form";
import Error from "./Error";
import Warning from "./Warning";
import Success from "./Success";

const Field = SemanticForm.Field;

class Form extends PureComponent {
    static propTypes = {
        error: PropTypes.string,
        warning: PropTypes.string,
        success: PropTypes.string
    };

    static defaultProps = {
        error: "",
        warning: "",
        success: ""
    };

    render() {
        const { children, error, warning, success, ...rest } = this.props;

        return (
            <SemanticForm {...rest}>
                {children}
                <Error error={error} touched={true} />
                <Warning warning={warning} touched={true} />
                <Success success={success} touched={true} />
            </SemanticForm>
        );
    }
}

export { Field };
export default Form;
