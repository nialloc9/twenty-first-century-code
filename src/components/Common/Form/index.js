import Error from "./Error";
import Information from "./Information";
import Input from "./Input";
import Select from "./Select";
import DatePicker from "./DatePicker";
import Checkbox from "./Checkbox";
import Label from "./Label";
import Success from "./Success";
import Warning from "./Warning";
import Form, { Field } from "./Form";
import rules from "./rules";

export {
    Error,
    Information,
    Input,
    Select,
    DatePicker,
    Checkbox,
    Label,
    Success,
    Warning,
    Form,
    Field,
    rules
};
