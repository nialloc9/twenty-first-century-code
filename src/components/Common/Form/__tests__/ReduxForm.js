/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import { shallow } from 'enzyme';

import Error from '../Error';
import Form from '../Form';
import Information from '../Information';
import Input from '../Input';
import Label from '../Label';
import Success from '../Success';
import TextArea from '../TextArea';
import Warning from '../Warning';
import rules from '../rules';

describe('Error', () => {
  test('renders', () => {
    const props = {
      error: 'test',
      touched: true,
    };

    const Component = shallow(<Error {...props} />);
    expect(Component).toMatchSnapshot();
  });
});

describe('Success', () => {
  test('renders', () => {
    const props = {
      success: 'test',
      touched: true,
    };

    const Component = shallow(<Success {...props} />);
    expect(Component).toMatchSnapshot();
  });
});

describe('Warning', () => {
  test('renders', () => {
    const props = {
      warning: 'test',
      touched: true,
    };

    const Component = shallow(<Warning {...props} />);
    expect(Component).toMatchSnapshot();
  });
});

describe('Form', () => {
  test('renders', () => {
    const props = {
      warning: null,
      success: null,
      error: null,
      onSubmit: jest.fn(),
    };

    const Component = shallow(<Form {...props} />);
    expect(Component).toMatchSnapshot();
  });
});

describe('Information', () => {
  test('renders', () => {
    const props = {
      information: 'test',
    };

    const Component = shallow(<Information {...props} />);
    expect(Component).toMatchSnapshot();
  });
});

describe('Input', () => {
  test('renders', () => {
    const props = {
      label: 'test',
      type: 'text',
      placeholder: 'test',
      input: {
        value: 'text',
        onChange: jest.fn(),
      },
      meta: {
        touched: false,
        warning: null,
        error: null,
      },
    };

    const Component = shallow(<Input {...props} />);
    expect(Component).toMatchSnapshot();
  });
});

describe('Label', () => {
  test('renders', () => {
    const props = {
      label: 'test',
    };

    const Component = shallow(<Label {...props} />);
    expect(Component).toMatchSnapshot();
  });
});

describe('TextArea', () => {
  test('renders', () => {
    const props = {
      label: 'test',
      meta: {
        touched: false,
        warning: null,
        error: null,
      },
      input: {
        value: 'text',
        onChange: jest.fn(),
      },
    };

    const Component = shallow(<TextArea {...props} />);
    expect(Component).toMatchSnapshot();
  });
});
