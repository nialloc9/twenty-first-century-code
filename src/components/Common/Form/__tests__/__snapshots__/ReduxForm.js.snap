// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error renders 1`] = `
<Styled(Message)
  margin="1.4286rem auto auto auto"
  negative={true}
>
  test
</Styled(Message)>
`;

exports[`Form renders 1`] = `
<Form
  as="form"
  onSubmit={[Function]}
>
  <Error
    error={null}
    touched={true}
  />
  <Warning
    touched={true}
    warning={null}
  />
  <Success
    success={null}
    touched={true}
  />
</Form>
`;

exports[`Information renders 1`] = `""`;

exports[`Input renders 1`] = `
<FormField>
  <Label
    inputLabel="test"
  />
  <Styled(Input)
    onChange={[Function]}
    placeholder="test"
    type="text"
    value="text"
  />
  <Error
    error={null}
    touched={false}
    warning={null}
  />
  <Warning
    error={null}
    touched={false}
    warning={null}
  />
</FormField>
`;

exports[`Label renders 1`] = `""`;

exports[`Success renders 1`] = `
<Styled(Message)
  margin="1.4286rem auto auto auto"
  positive={true}
  touched={true}
>
  test
</Styled(Message)>
`;

exports[`TextArea renders 1`] = `
<div>
  <Label
    label="test"
  />
  <TextArea
    as="textarea"
    onChange={[Function]}
    rows={3}
    value="text"
  />
  <Error
    error={null}
    touched={false}
    warning={null}
  />
  <Warning
    error={null}
    touched={false}
    warning={null}
  />
</div>
`;

exports[`Warning renders 1`] = `
<Styled(Message)
  color="yellow"
  margin="1.4286rem"
  touched={true}
>
  test
</Styled(Message)>
`;
